proto/docs/arbitration.md

```diff
@@ -1,74 +1,125 @@
-## Arbitration
-
-This server implementation of P4Runtime supports master-slave arbitration. In
-P4Runtime each client needs to open a bi-directional stream connection to the
-server using the `StreamChannel` RPC. The client should advertise its election
-id right away using a `MasterArbitrationUpdate` message. The process through
-which election ids are allocated to clients is out-of-scope of P4Runtime and of
-this document. The client with the highest election id is referred to as the
+## Controller Roles and Mastership Arbitration
+
+P4Runtime supports multiple controller roles to partition the control of the forwarding
+plane and P4 objects. For each role, P4Runtime supports master-arbitration across
+replicated instances of that controller (to protect against multi-master scenarios). At
+any given time, only one of the controller instances (for a given role) is considered a master.
+
+### Controller (or P4Runtime *client*) Role
+A given role is identified by `Role.id`.
+The `Role.config` describes the role configuration, i.e. what operations, P4 entities,
+behaviors, etc. are in the scope of a given role. If config is not set (default case),
+it implies that all P4 objects and control behaviors are in scope for that role
+(i.e. full pipeline access).
+
+The Role IDs are defined offline in agreement across the entire control plane.
+A client’s session with P4Runtime is tied to the role ID it passes over the stream
+channel (details below). The P4Runtime server will use `role` id for two purposes:
+- Arbitrate that each role has one and only one master client instance
+- Enforce the validity (access control) of P4Runtime requests based on the role config
+
+To keep the protocol simple and extensible, `config` is defined as a `protobuf.Any`,
+and concerete defintions are currently outside the scope of P4Runtime.
+
+### P4Runtime session establishment
+To establish a P4Runtime session, a client needs to open a bi-directional
+stream connection to the server using the `StreamChannel` RPC. The client should advertise
+its `role` and `election_id` right away using a `MasterArbitrationUpdate` message. The process
+through which election ids are allocated to clients is out-of-scope of P4Runtime and of
+this document. For a  given role, the client with the highest election id is referred to as the
 "master", while all other clients are referred to as "slaves".
 
 Only the master client can successfully:
 - perform `Write` requests
 - receive `PacketIn` messages
 - send `PacketOut` messages
 
-### `MasterArbitrationUpdate` client request
+### `MasterArbitrationUpdate` from client to server
 
-- if this is the first `MasterArbitrationUpdate` message sent by the client:
-  - if the election id is already used by another client connection, the server
+- If this is the first `MasterArbitrationUpdate` message sent by the client:
+  - If the election id is already used by another client connection, the server
```

Comment:
should the election id be unique across all roles or only within a given role id?

Reply:
had a discussion with this offline with @wmohsin; seems that we are leaning towards **unique for a given role id**; @wmohsin will update the spec to call this out explicitly

Reply:
Done